packages/nimble-components/src/table/specs/table-column-custom-sort-field-hld.md

```diff
@@ -0,0 +1,57 @@
+# Custom table column sort order HLD
+
+## Problem Statement
+
+In some cases, a client may have the need to specify a custom sort order for a column. For example, a table displaying SystemLink tags may have a field in their records for the stringified tag but may not want to sort that column by that stringified value. Instead, they want to provide their own sort order for tags that takes each tag's data type and value into account.
+
+## Links to Relevant Work Items and Reference Material
+
+[Azdo Feature 2665131 - Custom column sorting](https://dev.azure.com/ni/DevCentral/_workitems/edit/2665131)
+
+## Implementation / Design
+
+Columns that need support for a custom column sort order will extend their API to allow a client to optionally specify an additional field within the data that specifies the sort order of that column. This field is expected to be a numeric value that is the sorted data index of the record. When the table is sorted by this column, the table will use the `basic` sort operation and sort the records by the additionally specified field rather than the default `operandDataRecordFieldName` associated with the column.
+
+TanStack uses the same field internally for both sorting and grouping. Grouping by this custom sort field could likely lead to unexpected results since there isn't a guarantee that two values that render identically have the same sort sort index. Therefore, grouping and specifying a custom sort order will not be allowed on a column at the same time. If both are specified at the same time, the column will be considered invalid and the `invalidCustomSortWithGrouping` validation flag will be set on the column.
```

Comment:
This is possibly in the weeds / implementation detail, but what requirements and guarantees would we have for the values of the sort-by field?

1. Does it have to be an index (i.e a whole number from 0 to the number of records) or is any `Number` ok?
2. Is it ok for multiple records to have the same value? If so what is the tiebreaker to determine their sort order? The `field-name` / `label-field-name` display value? The original order of the data? No guarantees?
3. Do we need to define what happens if a value isn't provided (or is `null` / `NaN` / non-numeric, etc) for some records?

Reply:
The plan would be to use the existing `basicSortFunction`, which is used for our numeric-backed columns (e.g. number-text, duration). As a result:

1. Any `Number` is ok
2. There are no guarantees if you have multiple records with the same value. I'd guess the behavior is to maintain original order of the data, but we don't guarantee that (especially since that behavior would come from TanStack).
3. The sort function already has defined behavior for `null`, `undefined`, `NaN`. It also handles non-numeric cases, but it just wouldn't be locale-aware if you pass it strings. It uses the `>` operator for anything that isn't a special-cased value.

Feel free to look at the implementation of the sort function if you're curious:
https://github.com/ni/nimble/blob/d37ca77cda0729b42f05b4bac79be8ccc98ceb22/packages/nimble-components/src/table/models/sort-operations.ts#L50-L89